### PR TITLE
Beat `None` handling

### LINKED ISSUE #101: beat doesn't log when `None` was returned.
I noticed that questions -- that uses a beat -- which could return `None` as a value do not get logged. 

Here is what gets logged:
```
Actor examines thing
... hoping it's equal to <None>
    => <None>
```

Here is what I would expect to get logged:
```
Actor examines thing
    => <None>
... hoping it's equal to <None>
    => <None>
```

### PR DESCRIPTION
**Pull request by bandophahita**
Fixes https://github.com/ScreenPyHQ/screenpy/issues/101

### CODE REVIEW

File: screenpy/pacing.py [hunk 1]
```diff
@@ -87,7 +125,7 @@ def wrapper(*args: Any, **kwargs: Any) -> Any:
             completed_line = f"{line.format(actor, **cues)}"
             with the_narrator.stating_a_beat(func, completed_line, gravitas) as n_func:
                 retval = n_func(*args, **kwargs)
-                if retval is not None:
+                if retval is not None or function_attached_to_protocol(n_func):
```
Comment (perrygoy) [inline screenpy/pacing.py]: Does this work? `n_func` is the function that `beat` annotated wrapped in the Narrator's context(s); we could just do `func` here instead directly which is the function that `beat` is annotating.

File: tests/test_pacing.py [hunk 2]
```diff
@@ -18,6 +20,15 @@ def use(self):
         pass
 
 
+class NonesyQuestion(Answerable):
```
Comment (perrygoy) [inline tests/test_pacing.py]: Questions don't inherit from `Answerable`, so i'd prefer if we remove this so we can make it more real-like.

File: screenpy/pacing.py [hunk 3]
```diff
@@ -3,17 +3,55 @@
...
 from screenpy.narration import Narrator, StdOutAdapter
+from screenpy.protocols import Answerable, Resolvable
 
 Function = Callable[..., Any]
 the_narrator: Narrator = Narrator(adapters=[StdOutAdapter()])
 
 
+# credit to https://stackoverflow.com/a/25959545/2532408
+def get_class_that_defined_method(meth: Any) -> Optional[type]:
+    """attempt to identify what class a method/function came from"""
```
Comment (perrygoy) [inline screenpy/pacing.py]: I'd like a bit more of a descriptive docstring here and below on `function_attached_to_protocol`, but... as mentioned in other PRs, i'll be going through the docstrings shortly.

File: tests/test_pacing.py [hunk 4]
```diff
@@ -55,6 +66,18 @@ def test_interpolations(self, mocked_narrator) -> None:
...
+        See(NonesyQuestion(), IsEqualTo(None)).perform_as(Tester)
+
+        assert [r.msg for r in caplog.records] == [
+            "Tester sees if nonesyQuestion is equal to None.",
+            "    Tester examines NonesyQuestion",
+            "        => None",
+            "    ... hoping it's equal to None.",
+            "        => <None>",
+        ]
+
```
Comment (perrygoy) [inline tests/test_pacing.py:136]: I think we need one more test of `beat` not-logging a not-Answerable/Resolvable method. That is, a test that it won't log the return-value of `perform_as`, for example.

Also, if someone does something like this:
```python
class CornerCase:
    def answered_by(self, the_actor):
        ...
        
    @beat("Blah!")
    def do_not_log_me_man(self) -> None:
        # i just do something but don't want my return value to be logged
```
the code as-is will log `=> <None>` for a call to `.do_not_log_me_man()`, right? We might need to go a step further and compare the method name *and* the class against the protocol.

Comment (bandophahita) [reply tests/test_pacing.py:136]: We get that already in test_beat_logging.  It's just not immediately obvious.
```python
        assert [r.msg for r in caplog.records] == [
            "Tester sees if nonesyQuestion is equal to None.",
            "    Tester examines NonesyQuestion",
            "        => None",
            "    ... hoping it's equal to None.",
            "        => <None>",
        ]
        ]
```
If beat were to log when None was returned it would actually look like this:
```python
        assert [r.msg for r in caplog.records] == [
            "Tester sees if nonesyQuestion is equal to None.",
            "    Tester examines NonesyQuestion",
            "        => None",
            "    ... hoping it's equal to None.",
            "        => <None>",
            "    => <None>",  # <- we are proving this doesn't show up with the above version
        ]
```

File: tests/test_pacing.py (hunk 2 above)
Comment (bandophahita) [reply tests/test_pacing.py]: They don't in the primary library, but should they be prevented from doing so?  Inheriting ensures they take on whatever traits the protocol contain.

File: screenpy/pacing.py (hunk 1 above)
Comment (bandophahita) [reply screenpy/pacing.py]: It does work, but you're totally right it should be just `func` for clarity sake.

File: screenpy/pacing.py (hunk 3 above)
Comment (bandophahita) [reply screenpy/pacing.py]: Like more of an explanation of how it works?

File: tests/test_pacing.py (hunk 4 above)
Comment (bandophahita) [reply tests/test_pacing.py:136]: *coffee finally kicked in*  Oh, you're right!  I'm not sure how we handle that corner case.

File: tests/test_pacing.py (hunk 2 above)
Comment (bandophahita) [reply tests/test_pacing.py]: fixed

File: tests/test_pacing.py (hunk 4 above)
Comment (bandophahita) [reply tests/test_pacing.py:136]: So we are back to the dilemma; how do we know if a function SHOULD return a None object that we want logged?

We made exceptions for `answered_by` and `resolve` by checking if we are currently attached to a class which conforms to the protocols `Answerable` and `Resolvable`.  Should we take it a step further and see if we are in methods with those specific names?  That rules out any functions that could legitimately return None we want logged as well.

I thought about trying to use the annotation of the function.  If it annotates with ONLY `-> None` we skip logging, but if it's anything else we allow it to log?   

It feels dirty to rely on annotations to dictate behavior, but I'm not sure how else to determine what should get logged.

Comment (bandophahita) [reply tests/test_pacing.py:136]: Say a user wanted to use beat on any function that returns.  Should THOSE get logged?

```python
@beat("does thing")
def foo() -> Any:
    return x
```
If so, limiting to protocols doesn't work here.

File: screenpy/pacing.py (hunk 3 above)
Comment (perrygoy) [reply screenpy/pacing.py]: ... Honestly i can't remember now, sorry for the lack of information here!

I think it had started out as a comment about the missing `Args` section in the docstring, but then i realized we were missing them everywhere, so i edited this to just ask for a bit more description.

File: tests/test_pacing.py (hunk 4 above)
Comment (perrygoy) [reply tests/test_pacing.py:136]: >  If it annotates with ONLY -> None we skip logging, but if it's anything else we allow it to log?"

Oh, i kinda like that idea. That would skip us having to maintain a list of protocols which can/can't log and instead just use the annotation to see if we should be logging—yes, i do like that!

Comment (bandophahita) [reply tests/test_pacing.py:136]: I'm glad to hear you like it, cuz I can't shake the feeling I'm doing something very wrong.

Comment (perrygoy) [reply tests/test_pacing.py:136]: Is it that annotations aren't supposed to have runtime implications? Well we've crossed the Rubicon on that one 'cause we're already using Pydantic. :P

Comment (bandophahita) [reply tests/test_pacing.py:136]: Does this mean I should take this out of draft and officially submit it?

Comment (perrygoy) [reply tests/test_pacing.py:136]: Yes please! Let's get this reviewed and merged in, i like this approach.

File: screenpy/pacing.py [hunk 5]
```diff
@@ -3,17 +3,27 @@
...
 from typing import Any, Callable, Optional
 
 from screenpy.narration import Narrator, StdOutAdapter
 
 Function = Callable[..., Any]
+
 the_narrator: Narrator = Narrator(adapters=[StdOutAdapter()])
 
 
+def function_should_log_none(func: Function) -> bool:
```
Comment (perrygoy) [inline screenpy/pacing.py:19]: Should this be a private function? `_function_should_log_none`?

I'm guessing someone somewhere would probably like to use it, but we might want to signal that it's purpose is very narrow. Or maybe we don't? 🤔

**Review by perrygoy — DISMISSED** [review]
I like the tests, and i like the solution. I think it's elegant, if a little oddball. 👍 

Thanks again for finding this issue and fixing it, @bandophahita!

File: screenpy/pacing.py (hunk 5 above)
Comment (bandophahita) [reply screenpy/pacing.py:19]: I'm not worried about anyone using it since it's not in the main package `__init__`.  If they want to import it and use it... shouldn't be an issue for us.

**Comment by bandophahita** [comment]
Sorry to make you re-review.  Fixed merge conflict with 789b11c3 #109

**Review by perrygoy — APPROVED** [review]
Reapproved! 👍
